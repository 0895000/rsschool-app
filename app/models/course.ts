import { Document, Schema, model } from 'mongoose';

export interface ICourse {
    description: string;
    endDateTime: number;
    name: string;
    startDateTime: number;
    isActive: boolean;
}

export interface ICourseModel extends ICourse, Document {
    _id: string;
}

const CourseSchema: Schema = new Schema({
    _id: String,
    description: { type: String, default: '' },
    endDateTime: { type: Number, default: 0 },
    isActive: { type: Boolean, default: false },
    name: {
        default: '',
        required: true,
        type: String,
    },
    startDateTime: { type: Number, default: 0 },
});

const CourseModelName = 'Course';

export const CourseModel = model<ICourseModel>(CourseModelName, CourseSchema);
