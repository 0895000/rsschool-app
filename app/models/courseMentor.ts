import { Document, Schema, model } from 'mongoose';
import { IUserBase } from './user';

export interface ICouseUser {
    city: string;
    courseId: string;
    userId: string;
    excludeReason: string | undefined;
    isActive: boolean;
}

export interface ICourseMentor extends ICouseUser {
    menteeCapacity: number;
    mentees: Array<IUserBase>;
    preferedMentees: Array<IUserBase>;
}

export interface ICourseMentorModel extends ICourseMentor, Document {
    _id: string;
}

export const CourseMentorScheme: Schema = new Schema({
    city: String,
    courseId: String,
    excludeReason: { type: String, default: undefined },
    isActive: { type: Boolean, default: true },
    menteeCapacity: { type: Number, default: 5 },
    mentees: { type: [{ _id: { type: String } }], default: [] },
    preferedMentees: { type: [{ _id: { type: String } }], default: [] },
    userId: String,
});

export const CourseMentorModelName = 'CourseMentor';
export const CourseMentorDocument = model<ICourseMentorModel>(CourseMentorModelName, CourseMentorScheme);
