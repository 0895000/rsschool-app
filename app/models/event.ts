import { Document, Schema, model } from 'mongoose';

export interface IEventVenue {
    name: string;
    id: string;
}

export interface IEvent {
    id: string;
    courseId: string;
    description: string;
    endDateTime: number;
    name: string;
    startDateTime: number;
    venue: IEventVenue;
}

export interface IEventModel extends Document, IEvent {
    id: string;
}

export const EventVenueSchema: Schema = new Schema({
    name: String,
});

export const EventSchema: Schema = new Schema({
    courseId: String,
    description: String,
    endDateTime: Number,
    name: {
        required: true,
        type: String,
    },
    startDateTime: Number,
    venue: EventVenueSchema,
});

export const EventModelName = 'Event';
export const EventDocument = model<IEventModel>(EventModelName, EventSchema);
