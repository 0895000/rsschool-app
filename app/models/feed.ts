import { Schema, model, Document } from 'mongoose';

export enum FeedPriority {
    Critical = 0,
    High = 5,
    Medium = 10,
    Low = 15,
}

export interface IFeedRecord {
    priority?: FeedPriority;
    dateTime: number;
    actionType: Actions;
    data: any;
    courseId?: string;
    userId: string;
    entityType: FeedEntities;
}

export interface IFeedRecordModel extends Document, IFeedRecord {}

export const FeedRecordScheme: Schema = new Schema({
    actionType: String,
    courseId: String,
    data: Object,
    dateTime: Number,
    entityType: String,
    priority: { type: Number, default: 10 },
    userId: String,
});

export const FeedRecordModelName = 'FeedRecord';
export const FeedRecordDocument = model<IFeedRecordModel>(FeedRecordModelName, FeedRecordScheme);

export enum FeedEntities {
    Course = 'Course',
    Task = 'Task',
    User = 'User',
    Event = 'Event',
    Participation = 'Participation',
}

export enum FeedActions {
    'ENROLL' = 'ENROLL',
    'SIGNUP' = 'SIGNUP',
}

type CourseActions = FeedActions.ENROLL;
type UserActions = FeedActions.SIGNUP;

type Actions = CourseActions | UserActions;
