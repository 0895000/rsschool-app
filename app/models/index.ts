export * from './course';
export * from './event';

export interface IApiResponse<T> {
    data: T | T[] | null;
}
