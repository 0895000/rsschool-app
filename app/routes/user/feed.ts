import { NOT_FOUND, OK } from 'http-status-codes';
import * as Router from 'koa-router';
import { FeedRecordDocument } from '../../models';
import { userService } from '../../services';
import { setResponse } from '../utils';

async function getFeedRecords(userId: string) {
    return FeedRecordDocument.find({
        userId,
    })
        .sort({ dateTime: -1 })
        .exec();
}

export const getFeedRoute = async (ctx: Router.IRouterContext) => {
    const userId = ctx.state.user!._id;

    const user = await userService.getUserById(userId);
    if (user == null) {
        setResponse(ctx, NOT_FOUND);
        return;
    }
    const feedRecords = await getFeedRecords(userId);
    setResponse(ctx, OK, feedRecords);
};
