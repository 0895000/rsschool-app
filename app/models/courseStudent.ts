import { Document, Schema, model } from 'mongoose';
import { ICouseUser } from './courseMentor';
import { IUserBase } from './user';

export interface ICourseStudent extends ICouseUser {
    englishLevel: string;
    mentors: Array<IUserBase>;
}

export interface ICourseStudentModel extends ICourseStudent, Document {
    _id: string;
}

export const CourseStudentScheme: Schema = new Schema({
    city: String,
    courseId: String,
    englishLevel: { type: String, default: '' },
    excludeReason: { type: String, default: undefined },
    isActive: { type: Boolean, default: true },
    mentors: { type: [{ _id: { type: String } }], default: [] },
    userId: String,
});

export const CourseStudentModelName = 'CourseStudent';
export const CourseStudentDocument = model<ICourseStudentModel>(CourseStudentModelName, CourseStudentScheme);
