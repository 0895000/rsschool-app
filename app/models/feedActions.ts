import { FeedRecordModel, FeedActions, FeedEntities, FeedPriority } from './feed';

export function saveCourseEnrollFeedAction(userId: string, courseId: string, data: { text: string }) {
    return new FeedRecordModel({
        actionType: FeedActions.ENROLL,
        courseId,
        data,
        dateTime: Date.now(),
        entityType: FeedEntities.Course,
        priority: FeedPriority.High,
        userId,
    }).save();
}

export function saveUserSignupFeedAction(userId: string, data: { text: string }) {
    return new FeedRecordModel({
        actionType: FeedActions.SIGNUP,
        data,
        dateTime: Date.now(),
        entityType: FeedEntities.User,
        priority: FeedPriority.High,
        userId,
    }).save();
}
