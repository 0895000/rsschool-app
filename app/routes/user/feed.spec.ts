import mockingoose from 'mockingoose';
import { FeedActions, FeedEntities, IFeedRecordModel } from '../../models';
import { createTestContext } from '../../utils';
import { getFeedRoute } from './feed';

describe('User Feed route', () => {
    beforeEach(() => {
        mockingoose.resetAll();
    });

    xit('returns data', async () => {
        const ctx = createTestContext();
        const feedRecords = [
            {
                _id: '5b19146f65bbc725d2069082',
                actionType: FeedActions.ENROLL,
                courseId: 'course-id',
                data: {},
                dateTime: 0,
                entityType: FeedEntities.User,
                userId: 'apalchys',
            },
        ] as IFeedRecordModel[];

        mockingoose.User.toReturn({}, 'findOne');
        mockingoose.FeedRecord.toReturn(feedRecords, 'find');

        await getFeedRoute(ctx);
        expect(ctx.status).toBe(200);
        expect(ctx.body).toMatchSnapshot();
    });

    xit('returns 404 if no user', async () => {
        const ctx = createTestContext();
        mockingoose.User.toReturn(null, 'findOne');

        await getFeedRoute(ctx);
        expect(ctx.status).toBe(404);
        expect(ctx.body).toMatchSnapshot();
    });
});
