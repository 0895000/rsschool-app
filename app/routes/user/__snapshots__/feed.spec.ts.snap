// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Feed route returns 404 if no user 1`] = `
Object {
  "data": undefined,
}
`;

exports[`User Feed route returns data 1`] = `
Object {
  "data": Array [
    Object {
      "_id": "5b19146f65bbc725d2069082",
      "actionType": "ENROLL",
      "courseId": "course-id",
      "dateTime": 0,
      "entityType": "User",
      "priority": 10,
      "userId": "apalchys",
    },
  ],
}
`;
